## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [280]:
import seaborn as sns

titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
246,0,3,female,25.0,0,0,7.775,S,Third,woman,False,NaN,Southampton,no,True
2,1,3,female,26.0,0,0,7.925,S,Third,woman,False,NaN,Southampton,yes,True
819,0,3,male,10.0,3,2,27.900,S,Third,child,False,NaN,Southampton,no,False
551,0,2,male,27.0,0,0,26.000,S,Second,man,True,NaN,Southampton,no,True
363,0,3,male,35.0,0,0,7.050,S,Third,man,True,NaN,Southampton,no,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [281]:
from pandas import isna

amount_missed_td = isna(titanic_data).sum(axis="rows").to_frame()
amount_missed_td.T

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,0,0,177,0,0,0,2,0,0,0,688,2,0,0


### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [282]:
len_rows = len(titanic_data)
col_without_nan_td = titanic_data.dropna(axis="columns", thresh=len_rows / 2)
print(f"Numbers of rows: {len_rows}")
col_without_nan_td

Numbers of rows: 891


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


In [283]:
len_columns = len(col_without_nan_td.columns)
rows_without_nan_td = col_without_nan_td.dropna(axis="rows", thresh=len_columns / 2)
print(f"Numbers of columns: {len_columns}")
rows_without_nan_td

Numbers of columns: 14


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


In [284]:
titanic_data = rows_without_nan_td

from pandas import isna

amount_missed_td = isna(titanic_data).sum(axis="rows").to_frame()
amount_missed_td.T

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,0,0,177,0,0,0,2,0,0,0,2,0,0


### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [285]:
man_td = titanic_data[titanic_data["who"] == "man"]
woman_td = titanic_data[titanic_data["who"] == "woman"]
child_td = titanic_data[titanic_data["who"] == "child"]

avg_age_man_td = round(man_td["age"].mean())
avg_age_woman_td = round(woman_td["age"].mean())
avg_age_child_td = round(child_td["age"].mean())

avg_age_td = {"man": avg_age_man_td, "woman": avg_age_woman_td, "child": avg_age_child_td}


titanic_data["age"] = titanic_data["age"].fillna(titanic_data["who"].map(avg_age_td))
print(
    f"Average age of men: {avg_age_man_td}\nAverage age of women: {avg_age_woman_td}\nAverage age of children: {avg_age_child_td}"
)
titanic_data

Average age of men: 33
Average age of women: 32
Average age of children: 6


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,32.0,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [286]:
len_columns = len(titanic_data.columns)
clear_rows_td = titanic_data.dropna(axis="rows", thresh=len_columns - 1)
print(f"Numbers of rows: {len(titanic_data)}")
print(f"Numbers of rows after clear: {len(clear_rows_td)}")
titanic_data = clear_rows_td
titanic_data

Numbers of rows: 891
Numbers of rows after clear: 889


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,32.0,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [ ]:
towns_td = list(titanic_data["embark_town"].unique())
print(towns_td)

south_amount_td = (titanic_data["embark_town"] == "Southampton").sum()
cherb_amount_td = (titanic_data["embark_town"] == "Cherbourg").sum()
queen_amount_td = (titanic_data["embark_town"] == "Queenstown").sum()

passagers = {
    "Southampton": south_amount_td,
    "Cherbourg": cherb_amount_td,
    "Queenstown": queen_amount_td,
}
max(passagers)

['Southampton', 'Cherbourg', 'Queenstown']


'Southampton'

### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [288]:
survived_td = len(titanic_data[titanic_data["survived"] > 0])
survived_percent_td = round(survived_td / len(titanic_data) * 100, 2)
print(f"Percent of survived people: {survived_percent_td}")

Percent of survived people: 38.25


### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [289]:
is_survive = titanic_data["survived"] > 0
survived_people_td = titanic_data[is_survive]["embark_town"].value_counts()
print(type(survived_people_td))
survived_people_td

<class 'pandas.core.series.Series'>


embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: count, dtype: int64

### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [290]:
is_survive = titanic_data["survived"] > 0
surv_people_td = (
    titanic_data[is_survive]["class"].value_counts() / titanic_data["class"].value_counts() * 100
)
print(type(surv_people_td))
round(surv_people_td, 2)

<class 'pandas.core.series.Series'>


class
First     62.62
Second    47.28
Third     24.24
Name: count, dtype: float64

### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [291]:
is_rich = titanic_data["fare"] >= 100
is_survive = titanic_data["survived"] > 0
surv_rich_people = len(titanic_data[(is_rich) & (is_survive)]) / len(titanic_data[is_rich]) * 100
round(surv_rich_people, 2)

73.58

### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [279]:
is_child = titanic_data["who"] == "child"
is_alone = titanic_data["alone"] == "True"
child_alone_td = len(titanic_data[is_child & is_alone])
child_alone_td

0

Какие выводы вы можете сделать о выживших пассажирах Титаника? 

Дети одни не путешествуют, ну и большая часть богатых выжила, видимо, большинство было из Southampton - оттуда больше всего человек выжило